master/buildbot/reporters/telegram.py

```diff
@@ -0,0 +1,1055 @@
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+import io
+import json
+import random
+import shlex
+
+from twisted.internet import defer
+from twisted.internet import task
+from twisted.python import log
+from twisted.web import resource
+from twisted.web import server
+
+from buildbot import config
+from buildbot import util
+from buildbot.plugins.db import get_plugins
+from buildbot.process.results import CANCELLED
+from buildbot.process.results import EXCEPTION
+from buildbot.process.results import FAILURE
+from buildbot.process.results import RETRY
+from buildbot.process.results import SUCCESS
+from buildbot.process.results import WARNINGS
+from buildbot.reporters.words import Channel
+from buildbot.reporters.words import Contact
+from buildbot.reporters.words import StatusBot
+from buildbot.reporters.words import UsageError
+from buildbot.schedulers.forcesched import CollectedValidationError
+from buildbot.schedulers.forcesched import ForceScheduler
+from buildbot.util import bytes2unicode
+from buildbot.util import httpclientservice
+from buildbot.util import service
+from buildbot.util import unicode2bytes
+
+
+class TelegramChannel(Channel):
+
+    def __init__(self, bot, channel):
+        if isinstance(channel, dict):
+            super().__init__(bot, channel['id'])
+            self.chat = channel
+        else:
+            super().__init__(bot, channel)
+            self.chat = {'id': channel}
+
+    def __getitem__(self, item):
+        return self.chat[item]
+
+    def get(self, item, default=None):
+        return self.chat.get(item, default)
+
+    def update(self, src):
+        self.chat.update(src)
+
+    def list_notified_events(self):
+        if self.notify_events:
+            self.send("The following events are being notified:\n{}"
+                      .format("\n".join(sorted(
+                          "🔔 **{}**".format(n) for n in self.notify_events))))
+        else:
+            self.send("🔕 No events are being notified.")
+
+
+def collect_fields(fields):
+    for field in fields:
+        if field['fullName']:
+            yield field
+        if 'fields' in field:
+            yield from collect_fields(field['fields'])
+
+
+class TelegramContact(Contact):
+
+    def __init__(self, bot, user=None, channel=None):
+        super().__init__(bot, user, channel)
+        self.template = None
+
+    @property
+    def chatid(self):
+        return self.channel.id
+
+    @property
+    def userid(self):
+        if isinstance(self.user, dict):
+            return self.user['id']
+        else:
+            return self.user
+
+    @property
+    def user_full_name(self):
+        fullname = " ".join((self.user['first_name'],
```

Comment:
Could `first_name` be absent?

Reply:
No, it cannot. In [Telegram API](https://core.telegram.org/bots/api#user) it is not optional.